src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/TypesThatOwnDisposableFieldsShouldBeDisposable.cs

```diff
@@ -42,83 +42,78 @@ public override void Initialize(AnalysisContext context)
+        {
+            if (ctx.Operation is IAssignmentOperation { Target: IFieldReferenceOperation field } assignment
+                && IsObjectCreation(assignment.Value)
+                && !ctx.Options.IsConfiguredToSkipAnalysis(Rule, field.Field.Type, parent, ctx.Compilation)
+                && disposableFields.Contains(field.Field))
+            {
+                disposableFieldNamesBuilder.Add(field.Field.Name);
+            }
+            else if (ctx.Operation is IFieldInitializerOperation initializer && IsObjectCreation(initializer.Value))
+            {
+                var candidateFields = initializer.InitializedFields
+                    .Where(f => !ctx.Options.IsConfiguredToSkipAnalysis(Rule, f.Type, parent, ctx.Compilation))
+                    .Intersect(disposableFields);
+                foreach (var f in candidateFields)
                 {
-                    SemanticModel model = symbolContext.Compilation.GetSemanticModel(classDecl.SyntaxTree);
-                    IEnumerable<string> disposableFieldNames = classDecl.DescendantNodes(n => !(n is TTypeDeclarationSyntax) || ReferenceEquals(n, classDecl))
-                        .SelectMany(n => GetDisposableFieldCreations(n, model, disposableFieldsHashSet, symbolContext.CancellationToken))
-                        .Where(field => !symbolContext.Options.IsConfiguredToSkipAnalysis(Rule, field.Type, namedType, symbolContext.Compilation))
-                        .Select(field => field.Name);
-
-                    if (disposableFieldNames.Any())
-                    {
-                        // Type '{0}' owns disposable field(s) '{1}' but is not disposable
-                        symbolContext.ReportDiagnostic(
-                            namedType.CreateDiagnostic(Rule, namedType.Name, string.Join("', '", disposableFieldNames.Order())));
-                        return;
-                    }
+                    disposableFieldNamesBuilder.Add(f.Name);
                 }
             }
+        }
 
-            private static IEnumerable<TTypeDeclarationSyntax> GetClassDeclarationNodes(INamedTypeSymbol namedType, CancellationToken cancellationToken)
+        private static bool IsObjectCreation(IOperation op)
+        {
+            if (op is IConversionOperation conversion)
             {
-                foreach (SyntaxNode syntax in namedType.DeclaringSyntaxReferences.Select(s => s.GetSyntax(cancellationToken)))
-                {
-                    if (syntax != null)
-                    {
-                        TTypeDeclarationSyntax? classDecl = syntax.FirstAncestorOrSelf<TTypeDeclarationSyntax>(ascendOutOfTrivia: false);
-                        if (classDecl != null)
-                        {
-                            yield return classDecl;
-                        }
-                    }
-                }
+                return conversion.Operand.Kind == OperationKind.ObjectCreation;
```

Comment:
Consider using the below helper to walk down conversion operations: https://github.com/dotnet/roslyn-analyzers/blob/f356758767f3cb3b30a151788b834955bd03ecd7/src/Utilities/Compiler/Extensions/IOperationExtensions.cs#L706